oak/server/oak_runtime.cc

```diff
@@ -30,116 +30,113 @@
 
 namespace oak {
 
+namespace {
+// Name to use for the (sole) gRPC pseudo-Node.  This will not clash with any
+// dynamically created Node names because they are all of the form
+// "<config>-<number>".
+constexpr char kGrpcNodeName[] = "grpc";
+}  // namespace
+
 grpc::Status OakRuntime::Initialize(const ApplicationConfiguration& config) {
   LOG(INFO) << "Initializing Oak Runtime";
   if (!ValidApplicationConfig(config)) {
     return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Invalid configuration");
   }
   absl::MutexLock lock(&mu_);
 
-  // Accumulate the name => module_bytes map.
-  wasm_contents_.clear();
-  for (const auto& contents : config.wasm_contents()) {
-    wasm_contents_[contents.name()] = absl::make_unique<std::string>(contents.module_bytes());
-  }
-
-  // Create all of the nodes.  The validity check above will ensure there is at
-  // most one of each pseudo-Node type.
-  OakNode* log_node = nullptr;
-  for (const auto& node_config : config.nodes()) {
-    const std::string& node_name = node_config.node_name();
-    std::unique_ptr<OakNode> node;
-    if (node_config.has_web_assembly_node()) {
-      const auto& wasm_node = node_config.web_assembly_node();
-      LOG(INFO) << "Create Wasm node named {" << node_name << "}";
-      const std::string* module_bytes = wasm_contents_[wasm_node.wasm_contents_name()].get();
-      node = WasmNode::Create(this, node_config.node_name(), *module_bytes);
-    } else if (node_config.has_grpc_server_node()) {
-      LOG(INFO) << "Create gRPC pseudo-Node named {" << node_name << "}";
-      std::unique_ptr<OakGrpcNode> grpc_node = OakGrpcNode::Create(node_name);
-      grpc_node_ = grpc_node.get();  // borrowed copy
-      node = std::move(grpc_node);
-    } else if (node_config.has_log_node()) {
-      LOG(INFO) << "Create logging pseudo-node named {" << node_name << "}";
-      node = absl::make_unique<LoggingNode>(node_name);
-      log_node = node.get();
-    } else if (node_config.has_storage_node()) {
-      LOG(INFO) << "Created storage pseudo-Node named {" << node_name << "}";
-      node = absl::make_unique<StorageNode>(node_name, node_config.storage_node().address());
-    }
-    if (node == nullptr) {
-      return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Failed to create Oak Node");
+  // Accumulate the various data structures indexed by config name.
+  wasm_config_.clear();
```

Comment:
Can we call Initialize multiple times? (on the same instance)

Reply:
No, I think that would go horribly wrong.  Updated the oak_runtime.h comment.